src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/AuthUserRolesController.java

```
@@ -0,0 +1,122 @@
+            @ApiResponse(code = 401, message = "Unauthorized.", response = ErrorDetail.class),
+            @ApiResponse(code = 404, message = "User not found.", response = ErrorDetail.class)})
+    public ResponseEntity<Object> roles(@ApiParam(value = "The username of the user.", required = true) @PathVariable final String username) {
+        final var userOptional = userService.getAuthUserByUsername(username);
+
+        return userOptional.
+                map(u -> u.getAuthorities().stream().map(a -> new UserRole(a.getAuthorityName())).collect(Collectors.toSet())).
+                map(Object.class::cast).
+                map(ResponseEntity::ok).
+                orElse(notFoundResponse(username));
+    }
+
+    @PutMapping("/api/authuser/{username}/roles/{role}")
+    @PreAuthorize("hasRole('ROLE_MAINTAIN_OAUTH_USERS')")
+    @ApiOperation(value = "Add role to user.", notes = "Add role to user.", nickname = "addRole",
+            consumes = "application/json", produces = "application/json")
+    @ApiResponses(value = {
+            @ApiResponse(code = 204, message = "Created"),
+            @ApiResponse(code = 400, message = "Validation failed.", response = ErrorDetail.class),
+            @ApiResponse(code = 401, message = "Unauthorized.", response = ErrorDetail.class),
+            @ApiResponse(code = 404, message = "User not found.", response = ErrorDetail.class),
+            @ApiResponse(code = 409, message = "Role for user already exists.", response = ErrorDetail.class),
+            @ApiResponse(code = 500, message = "Server exception e.g. failed to insert row.", response = ErrorDetail.class)})
+    public ResponseEntity<Object> addRole(
+            @ApiParam(value = "The username of the user.", required = true) @PathVariable final String username,
+            @ApiParam(value = "The role to be added to the user.", required = true) @PathVariable final String role) {
+
+        final var userOptional = userService.getAuthUserByUsername(username);
```

Would this logic be better of in the service?

---

Tricky one.

The problem is that we return responses with different codes etc. depending on whether the user exists, whether the role exists etc.

One solution is to use magic to solve the problem, so the services throw exceptions of a certain type which then automatically get converted into the correct response code etc.  We haven't gone down that route here so there aren't any global exception handlers that cope with exceptions, instead the controller deals with the response manually.  It has the advantage that it is clear what is going on at all times, but it does lead to the controller having more logic

---

Well, you could throw the correct exceptions from your service, then handle them right here in this controller method.